dali/kernels/reduce/mean_stddev_gpu_impl.cuh

```diff
@@ -0,0 +1,386 @@
+template <typename Out, typename In, typename Acc =
+  default_sum_acc_t<Out, decltype(reductions::square()(In()))>>
+class RootMeanSquareImplGPU
+    : public ReduceImplGPU<Out, In, Acc, RootMeanSquareImplGPU<Out, In, Acc>>
+    , public RootMeanImplBase<Out, In, RootMeanSquareImplGPU<Out, In, Acc>> {
+ public:
+  using MeanBase = RootMeanImplBase<Out, In, RootMeanSquareImplGPU<Out, In, Acc>>;
+  using typename MeanBase::Postprocessor;
+
+  using Preprocessor = reductions::square;
+  template <int non_reduced_dims>
+  using PreprocessorBank = UniformPreprocessorBank<non_reduced_dims, Preprocessor>;
+
+  Preprocessor GetPreprocessorImpl() const { return {}; }
+  Preprocessor *GetPreprocessorsImpl(WorkArea &wa) const { return nullptr; }
+
+  template <int non_reduced_dims>
+  PreprocessorBank<non_reduced_dims> *
+  GetPreprocessorBanksImpl(WorkArea &wa, int axis, int_const<non_reduced_dims>) const {
+    return nullptr;
+  }
+
+  reductions::sum GetReduction() const { return {}; }
+};
+
+template <typename Out, typename In, typename Mean = Out, typename Acc = Out>
+class StdDevImplGPU : public ReduceImplGPU<Out, In, Acc, StdDevImplGPU<Out, In, Mean, Acc>>,
+                      public VarianceImplBase<Out, In, Mean, StdDevImplGPU<Out, In, Mean, Acc>>,
+                      public RootMeanImplBase<Out, In, StdDevImplGPU<Out, In, Mean, Acc>> {
+ public:
+  using ReduceBase = ReduceImplGPU<Out, In, Acc, StdDevImplGPU<Out, In, Mean, Acc>>;
+  using VarBase = VarianceImplBase<Out, In, Mean, StdDevImplGPU<Out, In, Mean, Acc>>;
+  using MeanBase = RootMeanImplBase<Out, In, StdDevImplGPU<Out, In, Mean, Acc>>;
+
+  using Preprocessor = typename VarBase::Preprocessor;
+  template <int non_reduced_dims>
+  using PreprocessorBank = typename VarBase::template PreprocessorBank<non_reduced_dims>;
+
+  using Postprocessor = typename MeanBase::Postprocessor;
+  reductions::sum GetReduction() const { return {}; }
+
+  void Run(KernelContext &kctx,
+           const OutListGPU<Out> &out,
+           const InListGPU<In> &in,
+           const InListGPU<Mean> &mean) {
+    this->InitMean(mean);
+    ReduceBase::Run(kctx, out, in);
+  }
+};
+
+template <typename Out, typename ScaleAndReg>
+struct RegularizedInvSqrt {
+  ScaleAndReg scale = 1, reg = 0;
+
+  template <typename T>
+  DALI_HOST_DEV Out operator()(T x) const {
+    float s = scale * x + reg;
+    return s ? ConvertSat<Out>(rsqrt(s)) : Out(0);
+  }
+};
+
+template <typename Out, typename In, typename Actual>
+class RegularizedInvRMS {
+ public:
+  Actual &This() { return static_cast<Actual&>(*this); }
+  const Actual &This() const { return static_cast<const Actual&>(*this); }
+
+  using param_t = std::conditional_t<std::is_same<Out, double>::value, double, float>;
+  using Postprocessor = RegularizedInvSqrt<Out, param_t>;
+
+  void SetRegularizationTerm(param_t reg) {
+    regularization_ = reg * reg;
+  }
+
+  param_t regularization_ = 0.0f;
+
+  Postprocessor *GetPostprocessorsImpl(WorkArea &wa) const {
```

Comment:
Just a thought, wouldn't one function like:
GetPostprocessor(int reduced_elements){} work as well. The ReduceImpl could call it in a loop for per-sample reductions, or once with TotalReducedElements() for batch-reduction and this would remove the need to define the same thing here twice, which you did for all of those clases in this file.

Reply:
No, it's per-sample and samples can have different reduction factors.